### ISSUE TITLE
Refactor ExpressionOptimizer::rewriteCast as a desugar rewriter

### ISSUE THREAD

**Issue by highker** [prestodb/presto#12829]
The optimization should happen within `PlanOptimizers`. The reason to specialize it in a later phase was due to lack of type in `Expression`. It should be moved back as a desugar rule.
blocked by #12828

**Comment by ajlab** [comment]
@highker Can I work on this?

**Comment by highker** [comment]
@ajlab sure! But this ticket is blocked by #12828 and #12546. Once these two migrations are done, feel free to send out the PR for this one.

**Comment by highker** [comment]
The method now got moved to `RowExpressionInterpreter::tryHandleCast`. But still needs clean up.

**Comment by fredlacis** [comment]
@highker Hi there! Can I work on this clean up? This would be my first issue.
